orchagent/switchorch.cpp

```diff
@@ -799,6 +1062,79 @@ void SwitchOrch::restartCheckReply(const string &op, const string &data, std::ve
     checkRestartReadyDone();
 }
 
+void SwitchOrch::onSwitchAsicSdkHealthEvent(sai_object_id_t switch_id,
+                                            sai_switch_asic_sdk_health_severity_t severity,
+                                            sai_timespec_t timestamp,
+                                            sai_switch_asic_sdk_health_category_t category,
+                                            sai_switch_health_data_t data,
+                                            const sai_u8_list_t &description)
+{
+    std::vector<swss::FieldValueTuple> values;
+    const string &severity_str = switch_asic_sdk_health_event_severity_reverse_map.at(severity);
+    const string &category_str = switch_asic_sdk_health_event_category_reverse_map.at(category);
+    string description_str;
+    const std::time_t &t = (std::time_t)timestamp.tv_sec;
+    stringstream time_ss;
+    time_ss << std::put_time(std::localtime(&t), "%Y-%m-%d %H:%M:%S");
+
+    switch (data.data_type)
+    {
+    case SAI_HEALTH_DATA_TYPE_GENERAL:
+    {
+        vector<uint8_t> description_with_terminator(description.list, description.list + description.count);
+        // Add the terminate character
+        description_with_terminator.push_back(0);
+        description_str = string(reinterpret_cast<char*>(description_with_terminator.data()));
+        // Remove unprintable characters but keep CR and NL
+        if (description_str.end() !=
+            description_str.erase(std::remove_if(
+                                  description_str.begin(),
+                                  description_str.end(),
+                                  [](unsigned char x) {
+                                      return (x != 0x0d) && (x != 0x0a) && !std::isprint(x);
+                                  }),
+                                  description_str.end()))
+        {
+            SWSS_LOG_NOTICE("Unprintable characters in description of ASIC/SDK health event");
+        }
+        break;
+    }
+    default:
+        SWSS_LOG_ERROR("Unknown data type %d when receiving ASIC/SDK health event", data.data_type);
+        // Do not return. The ASIC/SDK health event will still be recorded but without the description
+        break;
+    }
+
+    event_params_t params = {
+        { "sai_timestamp", time_ss.str() },
+        { "severity", severity_str },
+        { "category", category_str },
+        { "description", description_str }};
+
+    if (0 == gMyAsicName.size())
+    {
+        SWSS_LOG_NOTICE("[%s] ASIC/SDK health event occurred at %s, category %s: %s", severity_str.c_str(), time_ss.str().c_str(), category_str.c_str(), description_str.c_str());
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("[%s] ASIC/SDK health event occurred at %s, asic %s, category %s: %s", severity_str.c_str(), time_ss.str().c_str(), gMyAsicName.c_str(), category_str.c_str(), description_str.c_str());
```

Comment:
Could you list some examples of WARN/NOTICE SDK health notifications? May need to reduce the flood of these log messages if the notifications are not fatal as the NOTICE level is default for orchagent.

Reply:
There are some examples in the HLD
```
admin@sonic:~$ show asic-sdk-health-event received
Time                 Severity     Category   Description
-------------------  -----------  ---------  -----------------
2023-10-20 05:07:34  fatal        firmware   Command timeout
2023-10-20 03:06:25  fatal        software   SDK daemon keep alive failed
2023-10-20 05:07:34  fatal        asic_hw    Uncorrectable ECC error
2023-10-20 01:58:43  notice       asic_hw    Correctable ECC error
```

Reply:
We could update the SWSS log level similar to sai_switch_asic_sdk_health_category_t to avoid NOTICE SDK events when the SWSS log level is at FATAL.

Reply:
ok. Let's use `ERROR` for fatal events and `NOTICE` for other severities.